docs/spec/auth/subkeys_proposal.md

```diff
@@ -0,0 +1,156 @@
+
+```
+if stdSig.PubKeyIndex == 0 {
+  verify stdSig using acc.PubKey
+} else {
+  verify stdSig using acc.SubKeys[stdSig.PubKeyIndex - 1]
+}
+```
+
+In the AnteHandler, if a transaction comes from a SubKey, it verifies the msg route is permitted.
+
+```
+if msg.Route not in subKeyMetadata.PermissionedRoutes {
+  return ErrNotPermitted
+}
+```
+
+When a transaction signed by a SubKey is accepted and has paid fees, we need to log the fees it has paid.  We increase the `SubKeyMetadata.DailyFeeUsed` field.
+
+If a new transaction from a subkey comes in and the `tx.Fee + subKeyMetadata.DailyFeeUsed > subKeyMetadata.DailyFeeAllowance`, then the transaction is rejected as the subkey is trying to exceed its daily fee allowance.
+
+But now we need a way to decrease the `DailyFeeUsed` field once transactions are past the 24 hour window.  To do this we us a time based iterator similar to the ones used in the governance and staking queues.
+
+We start by defining a new struct called a `DailyFeeSpend`, which is what will be inserted into the Queue.
+
+```golang
+type DailyFeeSpend struct {
+  Address              sdk.AccAddress
+  SubKeyIndex          uint
+  FeeSpent             sdk.Coins
+}
+```
+
+When a transaction using a SubKey has paid its fee, we insert a new entry into this queue's store.
+- The key is `sdk.FormatTimeBytes(ctx.BlockTime.Add(time.Day * 1)`
+- The value is the Amino marshalled `DailyFeeSpend`
+
+Finally in the EndBlocker we iterate over the all the DailyFeeSpends in the queue which are expired (more than a day old).  We prune them from state, and deduct their FeeSpent amount from the corresponding SubKeyMetadata.
+
+```
+store := ctx.Store(spentFeeQueueStoreKey)
+iterator := sdk.Iterator(nil, sdk.FormatTimeBytes(ctx.BlockTime)) // pseudocode abstract this as a slice of DailyFeeSpend
+
+for _, dailyFeeSpend := range iterator {
+  acc := accountKeeper.GetAccount(dailyFeeSpend.Address)
+  acc.SubKeys[dailyFeeSpend.SubKeyIndex - 1].DailyFeeUsed -= dailyFeeSpend.FeeSpent
+  store.Delete(dailyFeeSpend)
+}
+```
+
+## Msgs
+
+In order to add, revoke, and update SubKeys, we would need to add some new `Msg` types to the auth module.
+
+We start by defining an `MsgAddSubKey` struct:
+
+```golang
+type MsgAddSubKey struct {
+  Address              sdk.AccAddress
+  PubKey               sdk.AccPubKey
+  PermissionedRoutes   []string
+  DailyFeeAllowance    sdk.Coins
+}
+```
+
+The handler of this Msg will create a new SubKeyMetadata struct using the data from this Msg, and append it to the end of the SubKeys slice of the Account associated with the Address.
```

Comment:
It should be noted that only a master key can add a new subkey, right?

Reply:
Technically, you can give any subkey access to the `MsgAddSubKey` route, but then you're essentially just giving it master control as well.  You should probably avoid doing that, but I don't think we should restrict it because there might be some use case I'm not thinking of yet.